aten/src/ATen/core/type_conversion.h

```diff
@@ -0,0 +1,217 @@
+  inline static int getTypeId() {
+    static const int id = LastTypeId++;
+    return id;
+  }
+
+  static std::atomic_int LastTypeId;
+  InternalMap m_map;
+};
+
+template <class ValueType>
+std::atomic_int TypeMap<ValueType>::LastTypeId(0);
+
+static TypeMap<c10::ClassTypePtr> tmap;
+
+namespace c10 {
+namespace detail {
+template <typename T>
+struct getTypePtr_ final {
+  static TypePtr call() {
+    auto res = tmap.find<T>();
+    if (res == tmap.end()) {
+        throw c10::Error("Trying to convert a class that's not registered.", "");
+    }
+    return std::dynamic_pointer_cast<Type>(res->second);
+  }
+//   static_assert(
+//       guts::false_t<T>::value,
+//       "Type could not be converted to any of the known types.");
+};
+
+template <>
+struct getTypePtr_<at::Tensor> final {
+  static TypePtr call() {
+    return TensorType::get();
+  }
+};
+template <>
+struct getTypePtr_<double> final {
+  static TypePtr call() {
+    return FloatType::get();
+  }
+};
+template <>
+struct getTypePtr_<int64_t> final {
+  static TypePtr call() {
+    return IntType::get();
+  }
+};
+template <>
+struct getTypePtr_<bool> final {
+  static TypePtr call() {
+    return BoolType::get();
+  }
+};
+template <>
+struct getTypePtr_<at::Scalar> final {
+  static TypePtr call() {
+    return NumberType::get();
+  }
+};
+template <>
+struct getTypePtr_<std::string> final {
+  static TypePtr call() {
+    return StringType::get();
+  }
+};
+template <class T>
+struct getTypePtr_<std::vector<T>> final {
+  static TypePtr call() {
+    static auto type = ListType::create(getTypePtr_<T>::call());
+    return type;
+  }
+};
+template <class T>
+struct getTypePtr_<c10::ArrayRef<T>> final {
+  static TypePtr call() {
+    static auto type = ListType::create(getTypePtr_<T>::call());
+    return type;
+  }
+};
+template <class T>
+struct getTypePtr_<c10::ListPtr<T>> final {
+  static TypePtr call() {
+    static auto type = ListType::create(getTypePtr_<T>::call());
+    return type;
+  }
+};
+template <class K, class V>
+struct getTypePtr_<std::unordered_map<K, V>> final {
+  static TypePtr call() {
+    static auto type =
+        DictType::create(getTypePtr_<K>::call(), getTypePtr_<V>::call());
+    return type;
+  }
+};
+template <class K, class V>
+struct getTypePtr_<c10::DictPtr<K, V>> final {
+  static TypePtr call() {
+    static auto type =
+        DictType::create(getTypePtr_<K>::call(), getTypePtr_<V>::call());
+    return type;
+  }
+};
+template <class T>
+struct getTypePtr_<at::optional<T>> final {
+  static TypePtr call() {
+    static auto type = OptionalType::create(getTypePtr_<T>::call());
+    return type;
+  }
+};
+} // namespace detail
+template <class T>
+inline TypePtr getTypePtr() {
+  // TODO: static_assert that a templated function exists, and throw a friendy
+  // error message if not
+  return detail::getTypePtr_<T>::call();
+}
+namespace ivalue {
+  #define CREATE_FACTORY_INIT(type) inline IValue from(type v) { return IValue(v); }
```

Comment:
Why have a `from` instead of using the constructor?

Reply:
We need a `from` function that can take in an arbitrary user specified type and turn that into an IValue. So the signature for the constructor would need to look like
```cpp
template<class T>
IValue(T x)
```

I talked with @smessmer about this, and it may have been impossible (https://stackoverflow.com/questions/3960849/c-template-constructor, although I can't quite remember if there was another conclusion), and is definitely considered a bad idea.